builder/replica_pipeline.cc

```diff
@@ -0,0 +1,609 @@
+    default: std::cerr << "Unknown intrinsic: " << op->intrinsic << std::endl; std::abort();
+    }
+    std::vector<std::string> args;
+    for (const auto& arg : op->args) {
+      args.push_back(print_expr(arg));
+    }
+    print_expr_assignment(call_name, "(", print_vector_elements(args), ")");
+  }
+
+  void visit(const let_stmt* op) override { fail("unimplemented let_stmt"); }
+  void visit(const block* op) override { fail("unimplemented block"); }
+  void visit(const loop* op) override { fail("unimplemented loop"); }
+  void visit(const call_stmt* op) override { fail("unimplemented call_stmt"); }
+  void visit(const copy_stmt* op) override { fail("unimplemented copy_stmt"); }
+  void visit(const allocate* op) override { fail("unimplemented allocate"); }
+  void visit(const make_buffer* op) override { fail("unimplemented make_buffer"); }
+  void visit(const clone_buffer* op) override { fail("unimplemented clone_buffer"); }
+  void visit(const crop_buffer* op) override { fail("unimplemented crop_buffer"); }
+  void visit(const crop_dim* op) override { fail("unimplemented crop_dim"); }
+  void visit(const slice_buffer* op) override { fail("unimplemented slice_buffer"); }
+  void visit(const slice_dim* op) override { fail("unimplemented slice_dim"); }
+  void visit(const truncate_rank* op) override { fail("unimplemented truncate_rank"); }
+  void visit(const check* op) override { fail("unimplemented check"); }
+
+  std::string print(const var& v) {
+    if (!v.defined()) {
+      return "var()";
+    }
+
+    auto it = vars_emitted_.find(v.sym());
+    if (it != vars_emitted_.end()) {
+      return it->second;
+    }
+
+    const auto& name = ctx_.name(v.sym());
+    vars_emitted_[v.sym()] = name;
+    return print_assignment_explicit(name, "var(ctx, \"", name, "\")");
+  }
+
+  std::string print(const std::vector<var>& vars) {
+    std::vector<std::string> var_names;
+    for (const auto& v : vars) {
+      var_names.push_back(print(v));
+    }
+    return print_vector(var_names);
+  }
+
+  std::string print(const std::vector<buffer_expr_ptr>& io) {
+    std::vector<std::string> bep_names;
+    for (const auto& bep : io) {
+      bep_names.push_back(print(bep));
+    }
+    return print_vector(bep_names);
+  }
+
+  std::string print(const buffer_expr_ptr& bep) {
+    const auto& name = ctx_.name(bep->sym());
+    if (buffers_emitted_.count(bep->sym())) {
+      return name;
+    }
+    buffers_emitted_.insert(bep->sym());
+
+    std::string size_code;
+    switch (bep->elem_size()) {
+    case 1: size_code = "sizeof(uint8_t)"; break;
+    case 2: size_code = "sizeof(uint16_t)"; break;
+    case 4: size_code = "sizeof(uint32_t)"; break;
+    case 8: size_code = "sizeof(uint64_t)"; break;
+    default: size_code = std::to_string(bep->elem_size()); break;
+    }
+
+    (void)print_assignment_explicit(name, "buffer_expr::make(ctx, \"", name, "\", ", size_code, ", ", bep->rank(), ")");
+
+    const auto matches = [](expr a, expr b) -> bool {
```

Comment:
This lambda is equivalent to just `match(a, b)`. It returns true if both are undef, false if only one is undef, and `match(a, b)` otherwise.